### PR TITLE
Microchip: MEC172x: kscan driver

### PR DESCRIPTION
**Pull request by jvasanth1**
Update keyscan driver to support MEC172x device

Signed-off-by: Jay Vasanth <jay.vasanth@microchip.com>

### CODE REVIEW

**Comment by jvasanth1** [comment]
Request to review this pull request.

File: drivers/kscan/Kconfig.xec
Comment (gmarull) [inline drivers/kscan/Kconfig.xec:89]: can't these be DT properties?

File: drivers/kscan/kscan_mchp_xec_v2.c
Comment (gmarull) [inline drivers/kscan/kscan_mchp_xec_v2.c:76]: please, don't use custom accessors, ref. https://github.com/zephyrproject-rtos/zephyr/pull/41918

File: drivers/kscan/kscan_mchp_xec_v2.c
Comment (gmarull) [inline drivers/kscan/kscan_mchp_xec_v2.c:1]: this seems to duplicate significant amount of code with "non-v2", why can't there be a common driver (or at least a shared codebase)?

**Review by gmarull — CHANGES_REQUESTED** [review]

File: drivers/kscan/kscan_mchp_xec_v2.c
Comment (jvasanth1) [reply drivers/kscan/kscan_mchp_xec_v2.c:76]: okay, will remove the custom accessors.

File: drivers/kscan/kscan_mchp_xec_v2.c
Comment (jvasanth1) [reply drivers/kscan/kscan_mchp_xec_v2.c:1]: okay, will evaluate

File: drivers/kscan/Kconfig.xec
Comment (jvasanth1) [reply drivers/kscan/Kconfig.xec:89]: These properties are dependent on the keypad connected to the platform. The hardware (kscan) needs to be configured based on that. Hence these properties are application dependent and hence in the Kconfig.

File: drivers/kscan/kscan_mchp_xec_v2.c
Comment (jvasanth1) [reply drivers/kscan/kscan_mchp_xec_v2.c:1]: This v2 version of driver is applicable to MEC172x device. Although the code is similar, we will be updating the v2 driver later to use pinctrl and there will be device tree specific changes. Also, we want to prioritize on the MEC172x device first.

File: drivers/kscan/Kconfig.xec
Comment (gmarull) [reply drivers/kscan/Kconfig.xec:89]: DT is the language used to describe hardware, so I still don't understand why DT can't be used for that.

File: drivers/kscan/kscan_mchp_xec_v2.c
Comment (gmarull) [reply drivers/kscan/kscan_mchp_xec_v2.c:1]: I've just run a diff between the 2 files and honestly, I think a generic driver has to be created, "v2" is duplicating too much code really. "v2" is also misleading.

**Review by gmarull — CHANGES_REQUESTED** [review]

File: drivers/kscan/Kconfig.xec
Comment (jvasanth1) [reply drivers/kscan/Kconfig.xec:89]: keypad is the external device connected to the platform, the kscan driver uses the Soc keyscan hardware to access the external keypad. Hence in DT we are describing keyscan hardware, but keypad details is unknown and depends on application usage and hence is in the kconfig. 

This is similar to [Kconfig.it8xxx2](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/kscan/Kconfig.it8xxx2)

Comment (gmarull) [reply drivers/kscan/Kconfig.xec:89]: Fair enough, thanks for the explanation. I guess it's ok then.

File: drivers/kscan/Kconfig.xec
Comment (tejlmand) [inline drivers/kscan/Kconfig.xec:60]: Why is the prompt a copy of the setting name, and not a real prompt text ?

File: drivers/kscan/Kconfig.xec
Comment (tejlmand) [inline drivers/kscan/Kconfig.xec:74]: Why is the prompt a copy of the setting name, and not a real prompt text ?

File: drivers/kscan/Kconfig.xec
Comment (tejlmand) [inline drivers/kscan/Kconfig.xec:67]: Why is the prompt a copy of the setting name, and not a real prompt text ?

File: drivers/kscan/Kconfig.xec
Comment (tejlmand) [inline drivers/kscan/Kconfig.xec:80]: Why is the prompt a copy of the setting name, and not a real prompt text ?

File: drivers/kscan/Kconfig.xec
Comment (tejlmand) [inline drivers/kscan/Kconfig.xec:86]: Why is the prompt a copy of the setting name, and not a real prompt text ?

**Review by tejlmand — CHANGES_REQUESTED** [review]

File: drivers/kscan/Kconfig.xec
Comment (jvasanth1) [reply drivers/kscan/Kconfig.xec]: updated with prompt text

File: drivers/kscan/Kconfig.xec
Comment (jvasanth1) [reply drivers/kscan/Kconfig.xec]: updated with prompt text

File: drivers/kscan/Kconfig.xec
Comment (jvasanth1) [reply drivers/kscan/Kconfig.xec]: updated with prompt text

File: drivers/kscan/Kconfig.xec
Comment (jvasanth1) [reply drivers/kscan/Kconfig.xec]: updated with prompt text

File: drivers/kscan/Kconfig.xec
Comment (jvasanth1) [reply drivers/kscan/Kconfig.xec]: updated with prompt text

File: drivers/kscan/kscan_mchp_xec_v2.c
Comment (jvasanth1) [reply drivers/kscan/kscan_mchp_xec_v2.c]: Added common driver instead of v2. 
Please review.

Thanks

File: drivers/kscan/kscan_mchp_xec_v2.c
Comment (jvasanth1) [reply drivers/kscan/kscan_mchp_xec_v2.c]: removed custom accessors

File: drivers/kscan/kscan_mchp_xec.c
Comment (gmarull) [inline drivers/kscan/kscan_mchp_xec.c:40]: You can store this directly:

```suggestion
	struct kscan_regs *regs;
```

Comment (jvasanth1) [reply drivers/kscan/kscan_mchp_xec.c]: Fixed in latest push

File: soc/arm/microchip_mec/mec172x/Kconfig.defconfig.mec172xnsz
Comment (gmarull) [inline soc/arm/microchip_mec/mec172x/Kconfig.defconfig.mec172xnsz:44]: driver can default based on compatible status instead (in driver's Kconfig):
```
default $(dt_compat_enabled,$(DT_COMPAT_XXXX))
```

File: drivers/kscan/kscan_mchp_xec.c
Comment (gmarull) [inline drivers/kscan/kscan_mchp_xec.c:492]: non-blocking naming suggestion:
```suggestion
static struct kscan_xec_data data;

static struct kscan_xec_config config = {
```

File: drivers/kscan/kscan_mchp_xec.c
Comment (keith-zephyr) [inline drivers/kscan/kscan_mchp_xec.c:81]: Non blocking suggestion: name this `void kscan_configure_girq(const struct device *dev, bool enable)`, or dropping the `enable` parameter and creating `kscan_enable_girq()` and `kscan_disable_girq()`.

**Review by keith-zephyr — DISMISSED** [review]

**Review by tejlmand — DISMISSED** [review]

File: drivers/kscan/kscan_mchp_xec.c
Comment (jvasanth1) [reply drivers/kscan/kscan_mchp_xec.c:492]: skipping this suggestion

File: soc/arm/microchip_mec/mec172x/Kconfig.defconfig.mec172xnsz
Comment (jvasanth1) [reply soc/arm/microchip_mec/mec172x/Kconfig.defconfig.mec172xnsz]: updated accordingly

**Review by gmarull — DISMISSED** [review]

File: drivers/kscan/kscan_mchp_xec.c
Comment (jvasanth1) [reply drivers/kscan/kscan_mchp_xec.c]: Changed to `void kscan_configure_girq(const struct device *dev, bool enable)` as per suggestion

**Review by keith-zephyr — APPROVED** [review]

**Comment by jvasanth1** [comment]
@gmarull could you approve again; i fixed one comment after your approval, hence need you to re-approve

**Review by gmarull — APPROVED** [review]

**Review by tejlmand — APPROVED** [review]

**Comment by jvasanth1** [comment]
Request to merge this PR

**Comment by mbolivar-nordic** [comment]
> Request to merge this PR

@jvasanth1 the merge will need to wait until after the merge window opens again, when v3.0 is released.

**Pull request merged**
